### PR TITLE
changes between 1.12.12 and 1.13.6 for backporting

### PR DESCRIPTION
**Pull request by dirk-thomas**
The following list of changes has been integrated into `ros_comm` 1.13.6 (Lunar) since the last Kinetic release (1.12.12).

**Backported:** (these changes are part of this PR)

* fix warn of --max-splits without --split #1237

  * bug fix

* fix documentation for cleanup_master_blacklist() #1253

  * documentation bug fix

* change error message for single threaded spinner #1164

  * helpful improvement, slightly modified patch

* make xmlrpcpp specific include directory work in devel space #1261

  * fix regression introduced in previous backport

* return an error status on error in rosbag #1257

  * bug fix

* replaced deprecated syntax using backticks #1259

  * helpful improvement

* performance improvement for lower/upper bound #1223

  * helpful improvement

* monotonic clock for callback queue timeouts #1250

  * helpful improvement

* topic subscription scalability fix in #1217

  * while risky (?) a significant improvement

* fix bad IPv6 initialization order #1268

  * bug fix

* fix incorrect docstrings in `rosnode/__init__.py` #1278

  * helpful improvement

* add decompress to free(state) before return #1313

  * bug fix

* raise the correct exception from AnyMsg.serialize #1311

  * bug fix

* make master process explicitly 'required' for parent launch #1228

  * helpful improvement, risk (?) of breaking if users were "relying" on launch to continue when the master crashes

* add TCP_INFO availability check #1211

  * bug fix

* add add_rostest_gmock function #1303

  * helpful improvement to be available for Kinetic and Lunar (as the underlying API in catkin)

* topic subscription scalability fix in #1281

  * while risky (?) a significant improvement

* add process listeners to XML RPC server #1319

  * bug fix

* fix xmlrpc timeout using monotonic clock #1249

  * helpful improvement

**Not backported:**

* tests for XmlRpcValue and XML conversion #1216

  * no need to backport

* unit tests and bug fixes for XmlRpcSocket #1218

  * no need to backport

* fix error handling #1226

  * only applicable to code on the Lunar branch

* unit tests and bug fixes for XmlRpcClient #1221

  * no need to backport

* unit tests for XmlRpcDispatch #1232

  * no need to backport

* update lz4 function name to support Trusty #1236

  * only applicable to code on the Lunar branch

* use namespaced macros instead of deprecated macros #1239

  * only applicable to code on the Lunar branch, the kinetic branch already contains a custom patch: 078481c0c113ae8c1c01947e2425e4f8a0ec62d0

* base64 tests #1242

  * no need to backport

* remove unnecessary xmlrpcpp dependency from message_filters #1264

  * no need to backport, also has the potential to break code which relies on this transitive dependency

* test and fix uninitialized data in XmlRpcClient #1244 

  * no need to backport

* tests and bug fixes for XmlRpcServer #1243

  * no need to backport

* add parameter to omit topics list from rosout logs #1234

  * new feature, not backported

* use roslaunch.core.printerrlog for printing error message #1193

  * no need to backport

* use SteadyTimer in message_filters #1247

  * changes behavior, not backported

* replace Thread.setDaemon() using new API #1276

  * no need to backport

* remove unreachable exceptions #1260

  * no need to backport

* printerrlog does not have file keyword #1317 

  * only applicable to code on the Lunar branch

* add --tcpnodelay TransportHint option to hz and delay commands #1296

  * new feature, not backported

* fix search strategy for python_logging config #1292

  * only applicable to code on the Lunar branch

* rename log macro argument from rate to period #1318

  * no need to backport

* remove preprended '|' from pretty-printed strings #1114

  * changes behavior, not backported

* pass through command-line args to the xmlloader when using the API #1115

  * new feature, not backported

@ros/ros_team @ros/ros_comm-maintainers Please comment on the decision which changes to (not) backport. Especially look at the changes marked with a "?".

### CODE REVIEW

**Comment by dirk-thomas** [comment]
For now #1249 was not included yet since it requires a release of `roscpp_core` first. I will add that commit next which is expected to fail CI. One I got feedback on this ticket I merge this (despite the CI failure) and do a release of `ros_comm` into Kinetic. After the rebuild of the Debian packages the devel job should pass again. :crossed_fingers:

**Review by wjwwood — APPROVED** [review]
lgtm, I consider the roslaunch master thing to be a bug and the two topic scaling fixes seem worth the risk and reasonably well acceptance tested.

**Comment by dirk-thomas** [comment]
@wjwwood Thank you for your feedback.

The latest Lunar release has just been synced. I will wait until early next week until I merge and release this into Kinetic.

**Pull request merged**
